core/src/util_normalize_table_generator.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+  Copyright:    © SIL International.
+  Description:  Generator for util_normalize_table.h
+  Create Date:  5 Jun 2024
+  Authors:      Steven R. Loomis
+
+  util_normalize_table.h is used under wasm by utilities in util_normalize.cpp to implement
+  normalization functions without needing ICU4C linked.
+
+  This generator is invoked automatically by meson as part of the build.
+*/
+
+#include "kmx/kmx_plus.h"
+#include "kmx/kmx_xstring.h"
+
+#define KMN_NO_ICU 0 // we will need ICU..
+
+#include "core_icu.h"
+
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <assert.h>
+
+#include <unicode/uchar.h>
+
+
+
+int
+write_nfd_table() {
+#ifndef __EMSCRIPTEN__
+  std::cerr << "Note: This is unusual - this generator is usually only run under emscripten!" << std::endl;
+#endif
+
+  // We write to stdout instead of to a file to avoid dealing with the filesystem under emscripten.
+
+  std::cerr << "Writing to stdout." << std::endl;
+
+  // write preamble
+  std::cout << "// GENERATED FILE: DO NOT EDIT" << std::endl;
+  std::cout << "//" << std::endl;
+  std::cout << "// util_normalize_table.h is generated by util_normalize_table_generator.cpp" << std::endl;
+  std::cout << "// and used by util_normalize.cpp" << std::endl;
+  std::cout << std::endl;
+  std::cout << "#pragma once" << std::endl;
+  std::cout << "#define KM_HASBOUNDARYBEFORE_UNICODE_VERSION \"" << U_UNICODE_VERSION << "\"" << std::endl;
+  std::cout << "#define KM_HASBOUNDARYBEFORE_ICU_VERSION \"" << U_ICU_VERSION << "\"" << std::endl;
+  std::cout << std::endl;
+  // we're going to need an NFD normalizer
+  UErrorCode status           = U_ZERO_ERROR;
+  const icu::Normalizer2 *nfd = icu::Normalizer2::getNFDInstance(status);
+  assert(U_SUCCESS(status));
+
+  // collect the raw list of chars that do NOT have a boundary before them.
+  std::vector<km_core_usv> noBoundary;
+  for (km_core_usv ch = 0; ch < 0x10FFFF; ch++) {
```

Comment:
```suggestion
  for (km_core_usv ch = 0; ch < km::core::kmx::Uni_MAX_CODEPOINT; ch++) {
```

Reply:
If we are starting at 0, why are we finishing at 0x10FFFE instead of 0x10FFFF? :grin: